<a href="https://colab.research.google.com/github/KarthikAmuthuri/Task-7/blob/main/Assignment10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Glass Classification


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
# Read train and test sets
train = pd.read_csv("trainKNN.txt", header=None)
train.columns = ['ID', 'RI', 'Na', 'Mg', 'Al', 'Si', 'K', 'Ca', 'Ba', 'Fe', 'Type of glass']
test = pd.read_csv("testKNN.txt", header=None)
test.columns = ['ID', 'RI', 'Na', 'Mg', 'Al', 'Si', 'K', 'Ca', 'Ba', 'Fe', 'Type of glass']

In [ ]:
train.head(10)

ID       RI     Na    Mg    Al     Si     K    Ca   Ba    Fe  Type of glass
0   1  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.00              1
1   2  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.00              1
2   3  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.00              1
3   4  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.00              1
4   5  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.00              1
5   6  1.51596  12.79  3.61  1.62  72.97  0.64  8.07  0.0  0.26              1
6   7  1.51743  13.30  3.60  1.14  73.09  0.58  8.17  0.0  0.00              1
7   8  1.51756  13.15  3.61  1.05  73.24  0.57  8.24  0.0  0.00              1
8   9  1.51918  14.04  3.58  1.37  72.08  0.56  8.30  0.0  0.00              1
9  10  1.51755  13.00  3.60  1.36  72.99  0.57  8.40  0.0  0.11              1

In [ ]:
test.head()

ID       RI     Na    Mg    Al     Si     K     Ca   Ba    Fe  \
0   68  1.52152  13.05  3.65  0.87  72.32  0.19   9.85  0.0  0.17   
1   69  1.52152  13.12  3.58  0.90  72.20  0.23   9.82  0.0  0.16   
2   70  1.52300  13.31  3.58  0.82  71.99  0.12  10.17  0.0  0.03   
3  144  1.51709  13.00  3.47  1.79  72.72  0.66   8.18  0.0  0.00   
4  145  1.51660  12.99  3.18  1.23  72.97  0.58   8.81  0.0  0.24   

   Type of glass  
0              1  
1              1  
2              1  
3              2  
4              2

from the above data we can drop the ID column from both the train and test datasets


In [ ]:
train = train.drop('ID',axis = 1)
test = test.drop('ID',axis = 1)

In [ ]:
train.describe()

RI          Na          Mg          Al          Si           K  \
count  196.000000  196.000000  196.000000  196.000000  196.000000  196.000000   
mean     1.518295   13.375204    2.758980    1.454337   72.635408    0.519388   
std      0.003055    0.783145    1.392641    0.491688    0.763578    0.672703   
min      1.511310   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516458   12.877500    2.362500    1.190000   72.317500    0.140000   
50%      1.517630   13.280000    3.480000    1.360000   72.810000    0.560000   
75%      1.518985   13.792500    3.610000    1.622500   73.080000    0.610000   
max      1.533930   15.790000    4.490000    3.500000   75.180000    6.210000   

               Ca          Ba          Fe  Type of glass  
count  196.000000  196.000000  196.000000     196.000000  
mean     8.910714    0.164235    0.050255       2.668367  
std      1.421490    0.485198    0.086359       2.062416  
min      5.430000    0.000000    0.000000       1.000000  
25%      8.220000    0.000000    0.000000       1.000000  
50%      8.575000    0.000000    0.000000       2.000000  
75%      9.092500    0.000000    0.090000       3.000000  
max     16.190000    3.150000    0.340000       7.000000

In [ ]:
test.describe()

RI         Na         Mg         Al         Si          K  \
count  18.000000  18.000000  18.000000  18.000000  18.000000  18.000000   
mean    1.519131  13.763333   1.873889   1.342222  72.820000   0.253889   
std     0.002791   1.083491   1.749753   0.581312   0.892004   0.265133   
min     1.511150  12.850000   0.000000   0.340000  71.360000   0.000000   
25%     1.517412  13.012500   0.000000   0.877500  72.212500   0.000000   
50%     1.519515  13.355000   2.395000   1.320000  72.685000   0.175000   
75%     1.521055  14.220000   3.570000   1.902500  73.382500   0.502500   
max     1.523000  17.380000   3.780000   2.170000  75.410000   0.760000   

              Ca         Ba         Fe  Type of glass  
count  18.000000  18.000000  18.000000      18.000000  
mean    9.460556   0.292778   0.130556       4.000000  
std     1.380432   0.617422   0.164798       2.222876  
min     6.650000   0.000000   0.000000       1.000000  
25%     8.635000   0.000000   0.000000       2.000000  
50%     9.065000   0.000000   0.015000       4.000000  
75%    10.090000   0.112500   0.240000       6.000000  
max    12.500000   1.670000   0.510000       7.000000

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196 entries, 0 to 195
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   RI             196 non-null    float64
 1   Na             196 non-null    float64
 2   Mg             196 non-null    float64
 3   Al             196 non-null    float64
 4   Si             196 non-null    float64
 5   K              196 non-null    float64
 6   Ca             196 non-null    float64
 7   Ba             196 non-null    float64
 8   Fe             196 non-null    float64
 9   Type of glass  196 non-null    int64  
dtypes: float64(9), int64(1)
memory usage: 15.4 KB


In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18 entries, 0 to 17
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   RI             18 non-null     float64
 1   Na             18 non-null     float64
 2   Mg             18 non-null     float64
 3   Al             18 non-null     float64
 4   Si             18 non-null     float64
 5   K              18 non-null     float64
 6   Ca             18 non-null     float64
 7   Ba             18 non-null     float64
 8   Fe             18 non-null     float64
 9   Type of glass  18 non-null     int64  
dtypes: float64(9), int64(1)
memory usage: 1.5 KB


standardizing the columns in both datasets expect the target column

In [ ]:
def standardize (df):
    for col in df.columns:
        if col != "Type of glass": # Don't standardize the categories
            df[col] = (df[col] - df[col].mean())/df[col].std()
    return df

In [ ]:
# Standardize data
train = standardize(train)
test = standardize(test)

In [ ]:
train.head()

RI        Na        Mg        Al        Si         K        Ca  \
0  0.888537  0.338119  1.242977 -0.720654 -1.120263 -0.682898 -0.113060   
1 -0.224221  0.657345  0.603903 -0.191863  0.123880 -0.058551 -0.760269   
2 -0.692234  0.197659  0.568000  0.174223  0.464382 -0.192340 -0.795443   
3 -0.207857 -0.210950  0.668529 -0.334230 -0.033275  0.075237 -0.485909   
4 -0.286405 -0.134335  0.618265 -0.435920  0.582248  0.045506 -0.591432   

        Ba        Fe  Type of glass  
0 -0.33849 -0.581932              1  
1 -0.33849 -0.581932              1  
2 -0.33849 -0.581932              1  
3 -0.33849 -0.581932              1  
4 -0.33849 -0.581932              1

splitting the data in X_train,X_test,y_train,y_test

In [ ]:
#splitting train data
y_train = train['Type of glass']
X_train = train.drop('Type of glass',axis =1)

#splitting test data
y_test = test['Type of glass']
X_test = test.drop('Type of glass',axis =1)

Building the models:

1. Square Euclidean Distance Model

2. Manhattan Distance Model

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
from scipy.spatial import distance
knc_euclidean_model = KNeighborsClassifier(n_neighbors = 8, metric = distance.sqeuclidean)# Square Euclidean distance model
knc_manhattan_model = KNeighborsClassifier(n_neighbors = 8, metric = distance.cityblock )# Manhattan distance model

Training the models

In [ ]:
knc_euclidean_model.fit(X_train,y_train)
knc_manhattan_model.fit(X_train,y_train)

KNeighborsClassifier(metric=<function cityblock at 0x7ff88f900290>,
                     n_neighbors=8)

In [ ]:
# Make predictions
euclid_predictions = knc_euclidean_model.predict(X_test) 
manhattan_predictions = knc_manhattan_model.predict(X_test)
df = pd.DataFrame({'actual': y_test, 'euclid': euclid_predictions, 'manhattan': manhattan_predictions})
df.head()

actual  euclid  manhattan
0       1       1          1
1       1       1          1
2       1       1          1
3       2       2          2
4       2       1          1

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
import warnings
warnings.filterwarnings('ignore') 

In [ ]:
# Evaluate performance
manhattan_count = len(df.loc[df['manhattan'] == df['actual']])
euclid_count = len(df.loc[df['euclid'] == df['actual']])
print('Manhattan Accuracy: {}%'.format(round(100*manhattan_count/len(df), 2)))
print(classification_report(y_test, manhattan_predictions, target_names=df['actual'].astype(str).unique()))
print ('\n')
print('Square Euclidean Accuracy: {}%'.format(round(100*euclid_count/len(df), 2)))
print(classification_report(y_test, euclid_predictions, target_names=df['actual'].astype(str).unique()))

Manhattan Accuracy: 66.67%
              precision    recall  f1-score   support

           1       0.50      1.00      0.67         3
           2       0.25      0.33      0.29         3
           3       0.00      0.00      0.00         3
           5       1.00      0.67      0.80         3
           6       1.00      1.00      1.00         3
           7       1.00      1.00      1.00         3

    accuracy                           0.67        18
   macro avg       0.62      0.67      0.63        18
weighted avg       0.62      0.67      0.63        18



Square Euclidean Accuracy: 61.11%
              precision    recall  f1-score   support

           1       0.60      1.00      0.75         3
           2       0.33      0.67      0.44         3
           3       0.00      0.00      0.00         3
           5       1.00      0.67      0.80         3
           6       1.00      0.33      0.50         3
           7       0.75      1.00      0.86         3

    accuracy  

**The performance of Manhattan Model Accuracy: 66.67% is good compare with the performance of Square Euclidean Model Accuracy: 61.11%.**